util/reflect.go

```diff
@@ -367,61 +413,368 @@ type NodeInfo struct {
-			nn := *ni
-			nn.FieldValue = ni.FieldValue.Index(i)
-			errs = AppendErrs(errs, forEachFieldInternal(&nn, in, out, iterFunction))
+	case IsTypeSlice(t):
+		schema := *(ni.Schema)
+		schema.ListAttr = nil
+		var pp []string
+		// Leaf-list elements share the parent schema.
+		if !schema.IsLeafList() {
+			pp = []string{schema.Name}
+		}
+		if IsNilOrInvalidValue(v) {
+			// Traverse the type tree only from this point.
+			nn := &NodeInfo{
+				Parent:         ni,
+				PathFromParent: pp,
+				Schema:         &schema,
+				FieldValue:     reflect.Zero(t.Elem()),
+			}
+			errs = AppendErrs(errs, forEachFieldInternal(nn, in, out, iterFunction))
+		} else {
+			for i := 0; i < ni.FieldValue.Len(); i++ {
+				nn := *ni
+				// The schema for each element is the list schema minus the list
+				// attrs.
+				nn.Schema = &schema
+				nn.Parent = ni
+				nn.PathFromParent = pp
+				nn.FieldValue = ni.FieldValue.Index(i)
+				errs = AppendErrs(errs, forEachFieldInternal(&nn, in, out, iterFunction))
+			}
 		}
 
-	case IsValueMap(ni.FieldValue):
-		for _, key := range ni.FieldValue.MapKeys() {
-			nn := *ni
-			nn.FieldValue = ni.FieldValue.MapIndex(key)
-			nn.FieldKey = key
-			nn.FieldKeys = ni.FieldValue.MapKeys()
-			errs = AppendErrs(errs, forEachFieldInternal(&nn, in, out, iterFunction))
+	case IsTypeMap(t):
+		schema := *(ni.Schema)
+		schema.ListAttr = nil
+		if IsNilOrInvalidValue(v) {
+			nn := &NodeInfo{
+				Parent:         ni,
+				PathFromParent: []string{schema.Name},
+				Schema:         &schema,
+				FieldValue:     reflect.Zero(t.Elem()),
+			}
+			errs = AppendErrs(errs, forEachFieldInternal(nn, in, out, iterFunction))
+		} else {
+			for _, key := range ni.FieldValue.MapKeys() {
+				nn := *ni
+				nn.Schema = &schema
+				nn.Parent = ni
+				nn.PathFromParent = []string{schema.Name}
+				nn.FieldValue = ni.FieldValue.MapIndex(key)
+				nn.FieldKey = key
+				nn.FieldKeys = ni.FieldValue.MapKeys()
+				errs = AppendErrs(errs, forEachFieldInternal(&nn, in, out, iterFunction))
+			}
 		}
 	}
 
-	return nil
+	return errs
+}
+
+// GetNodes returns the nodes in the data tree at the indicated path, relative
+// to the supplied root. If for a list node is missing, all values in the list
+// are returned. If the key is partial, all nodes matching the values present
+// in the key are returned.
+// If the root is the tree root, the path may be absolute.
+// GetNodes returns an error if the path is not found in the tree, or an element
+// along the path is nil.
+func GetNodes(schema *yang.Entry, root interface{}, path *gpb.Path) ([]interface{}, []*yang.Entry, error) {
+	return getNodesInternal(schema, root, path)
+}
+
+// getNodesInternal is the internal implementation of GetNode. In addition to
+// GetNode functionality, it can accept non GoStruct types e.g. map for a keyed
+// list, or a leaf.
+func getNodesInternal(schema *yang.Entry, root interface{}, path *gpb.Path) ([]interface{}, []*yang.Entry, error) {
```

Comment:
Suggest to add something to the function comment that tells us what the return values are -- I guess the `[]interface{}` is the set of values that the path matches, and `[]*yang.Entry` is the schema(s).

See the comment below about `leafref`. It seems to me that you can return a single value and state that it is a `leaf-list` even though the input schema is a `leaf` of type `leafref`. This might mean that marshalling at the client is wrong (i.e., it should not marshal to a JSON array, but a scalar).

Reply:
Per discussion, I've updated the comment and removed resolving the leafref. For the case of returning a leaf-list, the list itself is considered a single node and the caller must be prepared to iterate through it. The schema at the same index will be the leaf-ref schema (which is shared with the leaf-list entries).